File: apps/contextractor/src/handler.py
"""Request handler for content extraction."""

from __future__ import annotations

import hashlib
from datetime import datetime, timezone
from typing import Any

from apify import Actor
from crawlee.crawlers import PlaywrightCrawlingContext

from contextractor_engine import ContentExtractor, TrafilaturaConfig

from .extraction import (
    compute_content_info,
    extract_format,
    extract_metadata,
    save_content_to_kvs,
)


class ResultsCounter:
    """Thread-safe counter for tracking results."""

    def __init__(self, max_results: int) -> None:
        self.max_results = max_results
        self.count = 0

    def increment(self) -> int:
        """Increment and return new count."""
        self.count += 1
        return self.count

    def is_limit_reached(self) -> bool:
        """Check if max results limit is reached."""
        return self.max_results > 0 and self.count >= self.max_results


def create_request_handler(
    kvs: Any,
    dataset: Any | None,
    results_counter: ResultsCounter,
    browser_log_enabled: bool,
):
    """Create a request handler function.

    Args:
        kvs: Key-value store for content.
        dataset: Optional named dataset.
        results_counter: Counter for tracking results.
        browser_log_enabled: Whether to log browser console.

    Returns:
        Async handler function for PlaywrightCrawler.
    """

    async def handler(context: PlaywrightCrawlingContext) -> None:
        """Process a single page and extract content."""
        # Check if max results reached
        if results_counter.is_limit_reached():
            Actor.log.info(f'Max results ({results_counter.max_results}) reached, stopping')
            return

        url = context.request.url
        Actor.log.info(f'Processing {url}')

        # Enable browser console logging if requested
        if browser_log_enabled:
            context.page.on(
                'console',
                lambda msg: Actor.log.info(f'[Browser] {msg.type}: {msg.text}'),
            )

        html = await context.page.content()
        key_base = hashlib.md5(url.encode()).hexdigest()[:16]

        handler_config = context.request.user_data.get('config', {})

        # Build TrafilaturaConfig from raw dict
        trafilatura_config_raw = handler_config.get('trafilatura_config_raw', {})
        trafilatura_config = TrafilaturaConfig.from_json_dict(trafilatura_config_raw)

        extractor = ContentExtractor(config=trafilatura_config)

        # Build raw HTML info
        html_bytes = html.encode('utf-8')
        raw_html_info = compute_content_info(html_bytes)

        if handler_config.get('save_raw_html'):
            html_key = f'{key_base}-raw.html'
            await kvs.set_value(html_key, html, content_type='text/html; charset=utf-8')
            raw_html_info['key'] = html_key
            raw_html_info['url'] = await kvs.get_public_url(html_key)

        # Extract metadata using ContentExtractor
        metadata = extract_metadata(html, url, extractor)

        # Build dataset entry
        data: dict[str, Any] = {
            'loadedUrl': url,
            'rawHtml': raw_html_info,
            'loadedAt': datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'),
            'metadata': metadata,
            'httpStatus': 200,
        }

        # Save extracted formats
        await _save_extracted_formats(kvs, key_base, html, url, extractor, handler_config, data)

        # Push data to dataset
        if dataset:
            await dataset.push_data(data)
        else:
            await context.push_data(data)

        # Increment results count
        results_counter.increment()

        # Stop crawler if max results reached
        if results_counter.is_limit_reached():
            Actor.log.info(
                f'Max results ({results_counter.max_results}) reached, stopping crawler'
            )
            raise SystemExit(0)

        # Enqueue links if linkSelector is set
        await _enqueue_links(context, handler_config)

    return handler


async def _save_extracted_formats(
    kvs: Any,
    key_base: str,
    html: str,
    url: str,
    extractor: ContentExtractor,
    config: dict[str, Any],
    data: dict[str, Any],
) -> None:
    """Save extracted content in requested formats.

    Args:
        kvs: Key-value store.
        key_base: Base key for storage.
        html: Raw HTML content.
        url: Source URL.
        extractor: ContentExtractor instance.
        config: Handler configuration.
        data: Data dict to update with results.
    """
    format_configs = [
        ('save_text', 'txt', 'extractedText', 'text/plain; charset=utf-8'),
        ('save_json', 'json', 'extractedJson', 'application/json; charset=utf-8'),
        ('save_markdown', 'markdown', 'extractedMarkdown', 'text/markdown; charset=utf-8'),
        ('save_xml', 'xml', 'extractedXml', 'application/xml; charset=utf-8'),
        ('save_xmltei', 'xmltei', 'extractedXmlTei', 'application/xml; charset=utf-8'),
    ]

    for config_key, output_format, data_key, content_type in format_configs:
        if config.get(config_key):
            content = extract_format(html, output_format, extractor, url=url)
            if content:
                ext = 'tei.xml' if output_format == 'xmltei' else output_format
                if output_format == 'markdown':
                    ext = 'md'
                key = f'{key_base}.{ext}'
                data[data_key] = await save_content_to_kvs(kvs, key, content, content_type)


async def _enqueue_links(
    context: PlaywrightCrawlingContext,
    config: dict[str, Any],
) -> None:
    """Enqueue links from the page if configured.

    Args:
        context: Crawling context.
        config: Handler configuration.
    """
    link_selector = config.get('link_selector', '')
    if not link_selector:
        return

    current_depth = context.request.user_data.get('depth', 0)
    max_depth = config.get('max_crawling_depth', 0)

    if max_depth != 0 and current_depth >= max_depth:
        return

    globs = config.get('globs', [])
    excludes = config.get('excludes', [])
    new_config = {**config}
    new_depth = current_depth + 1

    await context.enqueue_links(
        selector=link_selector,
        globs=[g.get('glob') for g in globs if g.get('glob')] if globs else None,
        exclude_globs=[e.get('glob') for e in excludes if e.get('glob')] if excludes else None,
        keep_url_fragments=config.get('keep_url_fragments', False),
        user_data={'config': new_config, 'depth': new_depth},
    )

File: apps/contextractor-apify/src/handler.ts
import { createHash } from 'node:crypto';
import { ContentExtractor } from '@contextractor/engine';
import type { Actor, Dataset } from 'apify';
import type { PlaywrightCrawlingContext } from 'crawlee';
import type { CrawlConfig, SaveFormat } from './config.js';
import {
  type ContentInfo,
  computeContentInfo,
  extractFormat,
  type KvsLike,
  projectMetadata,
  saveContentToKvs,
} from './extraction.js';

export class ResultsCounter {
  private count = 0;
  constructor(public readonly maxResults: number) {}
  increment(): number {
    this.count += 1;
    return this.count;
  }
  isLimitReached(): boolean {
    return this.maxResults > 0 && this.count >= this.maxResults;
  }
  get value(): number {
    return this.count;
  }
}

export interface HandlerDeps {
  kvs: KvsLike;
  dataset: Dataset | null;
  resultsCounter: ResultsCounter;
  browserLogEnabled: boolean;
}

const COOKIE_DISMISS_SCRIPT = `() => {
  if (window.Didomi) {
    try { window.Didomi.setUserAgreeToAll(); return; } catch {}
  }
  const onetrust = document.querySelector('#onetrust-accept-btn-handler');
  if (onetrust) { onetrust.click(); return; }
  const cookiebot = document.querySelector('#CybotCookiebotDialogBodyLevelButtonLevelOptinAllowAll');
  if (cookiebot) { cookiebot.click(); return; }
  const quantcast = document.querySelector('.qc-cmp2-summary-buttons button[mode="primary"]');
  if (quantcast) { quantcast.click(); return; }
  const selectors = [
    '[class*="cookie"] button', '[id*="cookie"] button',
    '[class*="consent"] button', '[id*="consent"] button',
    'button[class*="accept"]', 'button[id*="accept"]',
  ];
  for (const sel of selectors) {
    const btn = document.querySelector(sel);
    if (btn) { btn.click(); return; }
  }
}`;

export function createRequestHandler(
  deps: HandlerDeps,
): (context: PlaywrightCrawlingContext) => Promise<void> {
  const { kvs, dataset, resultsCounter, browserLogEnabled } = deps;

  return async function handler(context: PlaywrightCrawlingContext): Promise<void> {
    if (resultsCounter.isLimitReached()) {
      context.log.info(`Max results (${resultsCounter.maxResults}) reached, stopping.`);
      return;
    }

    const url = context.request.url;
    context.log.info(`Processing ${url}`);

    if (browserLogEnabled) {
      context.page.on('console', (msg) => {
        context.log.info(`[Browser] ${msg.type()}: ${msg.text()}`);
      });
    }

    const handlerConfig = (context.request.userData?.config as CrawlConfig) ?? null;
    if (!handlerConfig) {
      context.log.warning('No crawl config in user data; skipping.');
      return;
    }

    if (handlerConfig.closeCookieModals) {
      try {
        await context.page.evaluate(COOKIE_DISMISS_SCRIPT);
        await context.page.waitForTimeout(1000);
      } catch {
        // Best effort.
      }
    }

    if (handlerConfig.maxScrollHeightPixels > 0) {
      try {
        await context.page.evaluate(async (max: number) => {
          let scrolled = 0;
          while (scrolled < max) {
            window.scrollBy(0, 500);
            scrolled += 500;
            await new Promise((r) => setTimeout(r, 100));
          }
          window.scrollTo(0, 0);
        }, handlerConfig.maxScrollHeightPixels);
      } catch {
        // Best effort.
      }
    }

    const html = await context.page.content();
    const keyBase = createHash('md5').update(url).digest('hex').slice(0, 16);

    const extractor = new ContentExtractor(handlerConfig.trafilaturaConfig);

    const rawHtmlInfo: ContentInfo = computeContentInfo(html);
    if (handlerConfig.save.includes('html')) {
      const htmlKey = `${keyBase}-raw.html`;
      await kvs.setValue(htmlKey, html, { contentType: 'text/html; charset=utf-8' });
      rawHtmlInfo.key = htmlKey;
      if (kvs.getPublicUrl) rawHtmlInfo.url = await kvs.getPublicUrl(htmlKey);
    }

    const metadata = projectMetadata(html, url, extractor);

    const data: Record<string, unknown> = {
      loadedUrl: url,
      rawHtml: rawHtmlInfo,
      loadedAt: new Date().toISOString().replace(/\.\d+Z$/, 'Z'),
      metadata,
      httpStatus: 200,
    };

    await saveExtractedFormats(kvs, keyBase, html, url, extractor, handlerConfig, data);

    if (dataset) {
      await dataset.pushData(data);
    } else {
      await context.pushData(data);
    }

    resultsCounter.increment();

    if (resultsCounter.isLimitReached()) {
      context.log.info(`Max results (${resultsCounter.maxResults}) reached, stopping crawler.`);
      throw new Error('MAX_RESULTS_REACHED');
    }

    if (handlerConfig.linkSelector) {
      await enqueueLinks(context, handlerConfig);
    }
  };
}

interface SaveSpec {
  format: SaveFormat;
  outputFormat: 'text' | 'json' | 'markdown';
  dataKey: string;
  contentType: string;
  ext: string;
}

const SAVE_SPECS: readonly SaveSpec[] = [
  {
    format: 'text',
    outputFormat: 'text',
    dataKey: 'extractedText',
    contentType: 'text/plain; charset=utf-8',
    ext: 'txt',
  },
  {
    format: 'json',
    outputFormat: 'json',
    dataKey: 'extractedJson',
    contentType: 'application/json; charset=utf-8',
    ext: 'json',
  },
  {
    format: 'markdown',
    outputFormat: 'markdown',
    dataKey: 'extractedMarkdown',
    contentType: 'text/markdown; charset=utf-8',
    ext: 'md',
  },
];

async function saveExtractedFormats(
  kvs: KvsLike,
  keyBase: string,
  html: string,
  url: string,
  extractor: ContentExtractor,
  config: CrawlConfig,
  data: Record<string, unknown>,
): Promise<void> {
  for (const spec of SAVE_SPECS) {
    if (!config.save.includes(spec.format)) continue;
    const content = extractFormat(html, spec.outputFormat, extractor, url);
    if (!content) continue;
    const key = `${keyBase}.${spec.ext}`;
    data[spec.dataKey] = await saveContentToKvs(kvs, key, content, spec.contentType);
  }
}

async function enqueueLinks(
  context: PlaywrightCrawlingContext,
  config: CrawlConfig,
): Promise<void> {
  const currentDepth = (context.request.userData?.depth as number | undefined) ?? 0;
  if (config.maxCrawlingDepth !== 0 && currentDepth >= config.maxCrawlingDepth) {
    return;
  }
  const newDepth = currentDepth + 1;
  const globs = config.globs.map((g) => g.glob).filter(Boolean);
  const excludes = config.excludes.map((e) => e.glob).filter(Boolean);
  await context.enqueueLinks({
    selector: config.linkSelector,
    ...(globs.length > 0 ? { globs } : {}),
    ...(excludes.length > 0 ? { exclude: excludes } : {}),
    userData: { config, depth: newDepth },
    transformRequestFunction: (req) => {
      req.keepUrlFragment = config.keepUrlFragments;
      return req;
    },
  });
}

export type { Actor };
